dialogue-clients/src/main/java/com/palantir/dialogue/clients/DialogueDnsResolution.java

```diff
@@ -0,0 +1,193 @@
+import com.google.common.net.InetAddresses;
+import com.palantir.conjure.java.client.config.ClientConfiguration;
+import com.palantir.dialogue.clients.DialogueDnsDiscoveryMetrics.Resolution_Result;
+import com.palantir.dialogue.core.TargetUri;
+import com.palantir.logsafe.Safe;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.UnsafeArg;
+import com.palantir.logsafe.logger.SafeLogger;
+import com.palantir.logsafe.logger.SafeLoggerFactory;
+import com.palantir.refreshable.Refreshable;
+import com.palantir.refreshable.SettableRefreshable;
+import com.palantir.tritium.metrics.registry.TaggedMetricRegistry;
+import java.net.InetAddress;
+import java.net.Proxy;
+import java.net.ProxySelector;
+import java.net.URI;
+import java.net.UnknownHostException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.List;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+/** Internal utility functionality for handling dialogue DNS resolution. */
+final class DialogueDnsResolution {
+    private static final SafeLogger log = SafeLoggerFactory.get(DialogueDnsResolution.class);
+
+    // This shouldn't be a static singleton if we can avoid it.
+    private static final ScheduledExecutorService SINGLETON_SCHEDULER = Executors.newSingleThreadScheduledExecutor();
+
+    static RefreshableUris refreshableUris(String serviceName, ClientConfiguration configuration) {
+        return refreshableUris(
+                serviceName, configuration.taggedMetricRegistry(), configuration.uris(), configuration.proxy());
+    }
+
+    static RefreshableUris refreshableUris(
+            String channelName, TaggedMetricRegistry registry, List<String> uris, ProxySelector selector) {
+        // Metrics for total number of dns refresh tasks are running per channel. This way
+        // we can detect if one is removed unexpectedly.
+        DialogueDnsDiscoveryMetrics metrics = DialogueDnsDiscoveryMetrics.of(registry);
+        Timer successTimer = metrics.resolution()
+                .channelName(channelName)
+                .result(Resolution_Result.SUCCESS)
+                .build();
+        Timer failureTimer = metrics.resolution()
+                .channelName(channelName)
+                .result(Resolution_Result.FAILURE)
+                .build();
+        ImmutableList<TargetUri> initial = resolveUris(channelName, uris, selector, successTimer, failureTimer);
+        SettableRefreshable<ImmutableList<TargetUri>> refreshable = Refreshable.create(initial);
+        Counter scheduledTasks = metrics.scheduled(channelName);
+        ScheduledFuture<?> dnsRefreshFuture = SINGLETON_SCHEDULER.scheduleWithFixedDelay(
+                () -> {
+                    try {
+                        refreshable.update(resolveUris(channelName, uris, selector, successTimer, failureTimer));
```

Comment:
i'm pretty ignorant about how all of this wires up... how does the list of `uris` within this thread get updated when the client configuration on disk changes? or does that allocate an entirely new channel, causing this thread to eventually die when the old one is gc'd?

Reply:
The `ScheduledFuture` is encapsulated within the returned `RefreshableUris` object, so that when `RefreshableUris.stop()` is invoked, the scheduled task is removed.

That's handled here in the non-reloading case, but isn't fully implemented for the reloading case yet: https://github.com/palantir/dialogue/pull/2118/files/3be9eebaf03d51853105389ca18e6bc45db0465f#diff-6d1086f7aca5ef649647a82c52683f1d85aa8c3a9568e9b4510669f1e749fc89R98-R101

Reply:
ah i see, sorry i missed your other comment